src/mds/mdstypes.h

```diff
@@ -1035,6 +1041,17 @@ struct fnode_t {
   snapid_t snap_purged_thru;   // the max_last_destroy snapid we've been purged thru
   frag_info_t fragstat, accounted_fragstat;
   nest_info_t rstat, accounted_rstat;
+  utime_t rstat_dirty_from;
+
+  // "last_rstat_dirty_from" and "rstat_dirty_from_delivered" are
+  // used to judge whether rstat_dirty_from has been received by
+  // inode auth and whether the current rstat_dirty_from can be reset.
+  //
+  // These two fields won't be stored in the underlying storage system,
+  // so they can be set without pre_dirty
+  utime_t last_rstat_dirty_from;
+  bool rstat_dirty_from_delivered = false;
```

Comment:
why not set rstat_dirty_from to zero at the same time mds setting pf->accounted_rstat = ipf>rstat

Reply:
@ukernel Hi, zheng, thanks for your review

Reply:
@ukernel Hi, zheng, thanks for your review

The reason why I added a seperate message as a signal to put rstat_dirty_from to zero is that I thought there could be such cases as:

1. mds.a encodes its lock state and send it to mds.b
2. mds.b fails before it can store the lock state to the underlying storage
3. another rstat whose rstat_dirty_from is later than that was sent to mds.b come up at mds.a
4. if we set rstat_dirty_from to 0 at the time mds setting pf->accounted_rstat to pf_rstat, then when mds.b come up again, it would receive a later rstat_dirty_from, which means the ealier one is lost.

So to make the rstat_dirty_from reliable, I added that RSTATDELIVERED message.

Am I right? Thanks:-)

Reply:
at step 1 and 2, rstat_dirty_from is left untouched. 
a step 3, mds rstat_dirty_from should be unchanged too, because it's value is smaller.

furthermore, rstat is no dirty when accounted_rstat == rstat. what's the point to set rstat_dirty_from

Reply:
oh, i think i get your point. do you mean that the accounted_rstat's sync to rstat is protected by the wrlock of nestlock which wouldn't be taken if the inodes' replica is offline, so, when accounted_rstat == rstat, mds.b must have already done with the previous rstat. Right?

If so, i think i should correct that, thank you:)